rust/arrow/src/json/reader.rs

```diff
@@ -888,265 +855,362 @@ impl Decoder {
+                    .null_bit_buffer(bool_nulls.freeze())
+                    .build()
+            }
+            DataType::Int8 => self.read_primitive_list_values::<Int8Type>(rows),
+            DataType::Int16 => self.read_primitive_list_values::<Int16Type>(rows),
+            DataType::Int32 => self.read_primitive_list_values::<Int32Type>(rows),
+            DataType::Int64 => self.read_primitive_list_values::<Int64Type>(rows),
+            DataType::UInt8 => self.read_primitive_list_values::<UInt8Type>(rows),
+            DataType::UInt16 => self.read_primitive_list_values::<UInt16Type>(rows),
+            DataType::UInt32 => self.read_primitive_list_values::<UInt32Type>(rows),
+            DataType::UInt64 => self.read_primitive_list_values::<UInt64Type>(rows),
+            DataType::Float16 => {
+                return Err(ArrowError::JsonError("Float16 not supported".to_string()))
+            }
+            DataType::Float32 => self.read_primitive_list_values::<Float32Type>(rows),
+            DataType::Float64 => self.read_primitive_list_values::<Float64Type>(rows),
+            DataType::Timestamp(_, _)
+            | DataType::Date32(_)
+            | DataType::Date64(_)
+            | DataType::Time32(_)
+            | DataType::Time64(_) => {
+                return Err(ArrowError::JsonError(
+                    "Temporal types are not yet supported, see ARROW-4803".to_string(),
+                ))
+            }
+            DataType::Utf8 => {
+                StringArray::from_iter(flatten_json_string_values(rows).into_iter())
+                    .data()
+            }
+            DataType::LargeUtf8 => {
+                LargeStringArray::from_iter(flatten_json_string_values(rows).into_iter())
+                    .data()
+            }
+            DataType::List(field) => {
+                let child = self
+                    .build_nested_list_array::<i32>(&flatten_json_values(rows), field)?;
+                child.data()
+            }
+            DataType::LargeList(field) => {
+                let child = self
+                    .build_nested_list_array::<i64>(&flatten_json_values(rows), field)?;
+                child.data()
+            }
+            DataType::Struct(fields) => {
+                // extract list values, with non-lists converted to Value::Null
+                let len = rows.len();
+                let num_bytes = bit_util::ceil(len, 8);
+                let mut null_buffer =
+                    MutableBuffer::new(num_bytes).with_bitset(num_bytes, false);
+                let mut struct_index = 0;
+                let rows: Vec<Value> = rows
+                    .iter()
+                    .map(|row| {
+                        if let Value::Array(values) = row {
+                            values.iter().for_each(|_| {
+                                bit_util::set_bit(null_buffer.data_mut(), struct_index);
+                                struct_index += 1;
```

Comment:
Enumerate + could return iterator instead of Vec?

Reply:
I ended up being forced to use vec because I was getting panics from "Iterator must be sized". We use `ExactSizeIterator`, which doesn't work when flattening lists because the exact number of flattened values can't be determined.
I'm not pleased with having to collect then iterate, but it was the only thing that seemed to work.

I don't mind if you make some changes and push them onto my branch.

Reply:
Makes sense. I also played a bit with the code, but it's hard to change the code to return iterators. Maybe it makes sense to create / push to a vec instead in this case?